# QUESTION # 1

# IMPORT ZIP CODES TO EXPLORE AREAS IN AND AROUND TORONTO

In [26]:
import numpy as np 
import pandas as pd
import codecs
from bs4 import BeautifulSoup

!wget https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
print("Done")

--2019-09-10 11:50:18--  https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
Resolving en.wikipedia.org (en.wikipedia.org)... 91.198.174.192, 2620:0:862:ed1a::1
Connecting to en.wikipedia.org (en.wikipedia.org)|91.198.174.192|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 79904 (78K) [text/html]
Saving to: ‘List_of_postal_codes_of_Canada:_M.2’

100%[======================================>] 79,904       298KB/s   in 0.3s   

2019-09-10 11:50:19 (298 KB/s) - ‘List_of_postal_codes_of_Canada:_M.2’ saved [79904/79904]

Done


# SORT POSTCODES BY BOROUGH AND NEIGHBORHOOD 

In [27]:
f = codecs.open('List_of_postal_codes_of_Canada:_M', encoding='utf-8')
html = f.read()
soup = BeautifulSoup(html, 'html.parser')

df = (pd.read_html(str(soup.table))[0])

#ADD THE CODE TO DROP THE FIRST ROW HEADING
df.drop(axis=0, index=0, inplace=True)

#ADD THE COLLUM NAMES
column_names = ['Postcode', 'Borough', 'Neighborhood'] 
df.columns = column_names

#ELIMINATE UNASSIGNED BOROUGHS AND SHOW COMPLETE DF
df = df[df['Borough'] != "Not assigned"].reset_index(drop=True) 
df



Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
..       ...               ...   
181      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
186      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
190      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
192      M5X  Downtown Toronto   
193      M8X         Etobicoke   
194      M8X         Etobicoke   
195      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
199      M8Y         Etobicoke   
200      M8Y         Etobicoke   
201      M8Y         Etobicoke   
202      M8Y         Etobicoke   
203      M8Y         Etobicoke   
204      M8Y         Etobicoke   
205      M8Y         Etobicoke   
206      M8Z         Etobicoke   
207      M8Z         Etobicoke   
208      M8Z         Etobicoke   
209      M8Z         Etobicoke   
210      M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                                         Harbourfront  
3                                          Regent Park  
4                                     Lawrence Heights  
5                                       Lawrence Manor  
6                                         Not assigned  
7                                     Islington Avenue  
8                                                Rouge  
9                                              Malvern  
10                                     Don Mills North  
11                                    Woodbine Gardens  
12                                       Parkview Hill  
13                                             Ryerson  
14                                     Garden District  
15                                           Glencairn  
16                                          Cloverdale  
17                                           Islington  
18                                        Martin Grove  
19                                    Princess Gardens  
20                                     West Deane Park  
21                                      Highland Creek  
22                                          Rouge Hill  
23                                          Port Union  
24                                     Flemingdon Park  
25                                     Don Mills South  
26                                    Woodbine Heights  
27                                      St. James Town  
28                                  Humewood-Cedarvale  
29                                   Bloordale Gardens  
..                                                 ...  
181                                        Thistletown  
182           

In [28]:
na = df[df['Neighborhood']=="Not assigned"]
for index in range(len(na)):
    r = na.iloc[index]
    r['Neighborhood']=r['Borough']
    df.iloc[na.index[0]]['Neighborhood'] = r['Borough']

df.head 

<bound method NDFrame.head of     Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Down

# ELIMINATED UNASSIGNED ZIP CODES 

In [29]:
# combine the neighborhoods seperated by comma using join and drop method for any duplicates
aggregations = {
    'Neighborhood':lambda x:', '.join(x)}
g = df.groupby('Postcode').agg(aggregations).reset_index()
x = df.drop('Neighborhood',axis=1)
m = pd.merge(x, g, on='Postcode', how='left')
dd = m.drop_duplicates()
dd.to_csv("Toronto_Q1_Answer.csv", index=False)
dd


Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
..       ...               ...   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10                                     Don Mills North  
11                     Woodbine Gardens, Parkview Hill  
13                            Ryerson, Garden District  
15                                           Glencairn  
16   Cloverdale, Islington, Martin Grove, Princess ...  
21              Highland Creek, Rouge Hill, Port Union  
24                    Flemingdon Park, Don Mills South  
26                                    Woodbine Heights  
27                                      St. James Town  
28                                  Humewood-Cedarvale  
29   Bloordale Gardens, Eringate, Markland Wood, Ol...  
33                   Guildwood, Morningside, West Hill  
36                                         The Beaches  
37                                         Berczy Park  
38                                 Caledonia-Fairbanks  
39                                              Woburn  
40                                             Leaside  
41                                  Central Bay Street  
42                                            Christie  
43                                           Cedarbrae  
44                                   Hillcrest Village  
45     Bathurst Manor, Downsview North, Wilson Heights  
48                                    Thorncliffe Park  
..                                                 ...  
130                                 North Toronto West  
131           

# LISTING OF COMBINED ZIPS AND NEIGHBORHOODS

In [30]:
# shape method to print rows  
dd.shape

(103, 3)

# THERE ARE 103 ZIPS ABOVE SORTED INTO UNIQUE NEIGBORHOODS IN THE CITY OF TORONTO

# QUESTION # 2 

# DOWNLOADED CSV FILE FOR GEO LOCATE DATA FOR THE CITY OF TORONTO

In [32]:
#downloadcsv file
!wget http://cocl.us/Geospatial_data

--2019-09-10 11:51:01--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 169.48.113.201
Connecting to cocl.us (cocl.us)|169.48.113.201|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2019-09-10 11:51:01--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|169.48.113.201|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2019-09-10 11:51:04--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 107.152.26.197, 107.152.27.197
Connecting to ibm.box.com (ibm.box.com)|107.152.26.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2019-09-10 11:51:04--  https://ibm.box.com/public/static/9afz

In [33]:
# geocoder for LAT LONG data
geo = pd.read_csv('Geospatial_data')
geo.columns = ['Postcode', 'Latitude', 'Longitude']

df_outer = pd.merge(dd, geo, on='Postcode', how='inner')
df_outer.to_csv("Toronto_Q2_Answer.csv", index=False) 
df_outer

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
..       ...               ...   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Harbourfront, Regent Park  43.654260 -79.360636  
3                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                       Rouge, Malvern  43.806686 -79.194353  
7                                      Don Mills North  43.745906 -79.352188  
8                      Woodbine Gardens, Parkview Hill  43.706397 -79.309937  
9                             Ryerson, Garden District  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   Cloverdale, Islington, Martin Grove, Princess ...  43.650943 -79.554724  
12              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
13                    Flemingdon Park, Don Mills South  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Bloordale Gardens, Eringate, Markland Wood, Ol...  43.643515 -79.577201  
18                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                                 Caledonia-Fairbanks  43.689026 -79.453512  
22                                              Woburn  43.770992 -79.216917  

# SORTED LISTED ABOVE OF THE 103 ZIPS BY LAT AND LONG FOR CLUSTER MAPPING

In [34]:
CLIENT_ID = 'YLG2MFIW20HL3IC11TZ2N1PHFKFI3NW4HVPWU3EGO4Y2ADXO' 
CLIENT_SECRET = 'I3LQSZE3WIMLLQ5HVZY0ULIZL3E0QWAL5T14M2BLI1TKW4C0'
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)



Your credentails:
CLIENT_ID: YLG2MFIW20HL3IC11TZ2N1PHFKFI3NW4HVPWU3EGO4Y2ADXO
CLIENT_SECRET:I3LQSZE3WIMLLQ5HVZY0ULIZL3E0QWAL5T14M2BLI1TKW4C0


# USING FOUR SQUARE TO OBTAIN UNIQUE VENUES IN TORONTO

# QUESTION 3 MAP NEIGHBORHOOD EXPLORATION DATA AND MAP CLUSTER 

In [37]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    LIMIT = 75
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # make API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # Asking for retrun for relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

# function to obtain the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import json # library for JSON files


from geopy.geocoders import Nominatim # for latitude and longitude values

import requests # library for requests

# Matplotlib for map plotting
import matplotlib.cm as cm
import matplotlib.colors as colors

# k-means for clustering
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium 
print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [41]:
#43.628841 	-79.520999
latitude = 43.628841
longitude = -79.520999


# map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# markers for map
for lat, lng, borough, neighborhood in zip(df_outer['Latitude'], df_outer['Longitude'], df_outer['Borough'], df_outer['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto


## CLUSTER MAP SHOWING CLUSTER DENSITY WITHIN 500 MILES OF THE CITY OF TORONTO

In [47]:
toronto_data = df_outer.set_index('Borough').filter(regex=".*Toronto",axis=0)
toronto_data.head()


toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
The Danforth West, Riverdale
Design Exchange, Toronto Dominion Centre
Brockton, Exhibition Place, Parkdale Village
The Beaches West, India Bazaar
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Moore Park, Summerhill East
Chinatown, Grange Park, Kensington Market
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

## SORTED LISTING OF VENUE LOCATIONS IN TORONTO

In [39]:
print(toronto_venues.shape)
toronto_venues.head()
toronto_venues.groupby('Neighborhood').count()
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

(1719, 7)
There are 237 uniques categories.


## THERE ARE 237 UNIQUE CATEGORIES WITHIN THE CITY OF TORONTO

In [48]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# put neighborhood column back into dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# neighborhood column to the first column to define and sort data
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...                 0                 0      0   
1             0  ...                 0                 0      0   
2             0  ...                 0                 0      0   
3             0  ...                 0                 0      0   
4             0  ...                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  
0                      0         0          0            0  
1                      0         0          0            0  
2                      0         0          0            0  
3                      0         0          0            0  
4                      0         0          0            0  

[5 rows x 229 columns]

In [49]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped
toronto_grouped.shape

(38, 229)

In [42]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
         venue  freq
0  Coffee Shop  0.08
1         Café  0.05
2          Bar  0.04
3   Steakhouse  0.04
4   Restaurant  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2      Farmers Market  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Exhibition Place, Parkdale Village----
                   venue  freq
0         Breakfast Spot  0.09
1                   Café  0.09
2            Coffee Shop  0.09
3      Convenience Store  0.05
4  Performing Arts Venue  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.05
2          Restaurant  0.05
3          Smoke Shop  0.05
4         Pizza Place  0.05


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0   Airport Service  0.20
1  Airport Te

## LISTING OF THE TOP 5 VENUES IN TORONTO BY EACH NEIGHBORHOOD

In [50]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# columns for top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# make new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Steakhouse                  Café   American Restaurant   
1          Cocktail Bar            Steakhouse           Cheese Shop   
2                  Café        Breakfast Spot         Grocery Store   
3           Yoga Studio                  Park            Smoke Shop   
4      Airport Terminal       Harbor / Marina           Coffee Shop   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0                   Bar           Concert Hall                 Hotel   
1                  Café     Seafood Restaurant              Beer Bar   
2          Intersection  Performing Arts Venue             Pet Store   
3            Skate Park                Brewery         Burrito Place   
4      Sculpture Garden               Boutique                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Breakfast Spot            Restaurant           Burger Joint  
1                Bakery        Farmers Market         Breakfast Spot  
2          Climbing Gym  Caribbean Restaurant             Restaurant  
3               Butcher            Restaurant             Comic Shop  
4         Boat or Ferry        Airport Lounge     Airport Food Court

In [51]:
# add clusters defined as 5
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels for rows
kmeans.labels_[0:10] 



# cluster labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to include latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last collumns

Postcode               Neighborhood   Latitude  Longitude  \
Borough                                                                      
Downtown Toronto      M5A  Harbourfront, Regent Park  43.654260 -79.360636   
Downtown Toronto      M5B   Ryerson, Garden District  43.657162 -79.378937   
Downtown Toronto      M5C             St. James Town  43.651494 -79.375418   
East Toronto          M4E                The Beaches  43.676357 -79.293031   
Downtown Toronto      M5E                Berczy Park  43.644771 -79.373306   

                  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
Borough                                                                        
Downtown Toronto               0           Coffee Shop                  Park   
Downtown Toronto               0           Coffee Shop        Clothing Store   
Downtown Toronto               0           Coffee Shop                  Café   
East Toronto                   0                   Pub     Health Food Store   
Downtown Toronto               0           Coffee Shop          Cocktail Bar   

                      3rd Most Common Venue 4th Most Common Venue  \
Borough                                                             
Downtown Toronto                        Pub                Bakery   
Downtown Toronto  Middle Eastern Restaurant                  Café   
Downtown Toronto                 Restaurant                 Hotel   
East Toronto                          Trail  Other Great Outdoors   
Downtown Toronto                 Steakhouse           Cheese Shop   

                 5th Most Common Venue 6th Most Common Venue  \
Borough                                                        
Downtown Toronto                  Café            Restaurant   
Downtown Toronto        Cosmetics Shop             Bookstore   
Downtown Toronto        Cosmetics Shop             Gastropub   
East Toronto               Wings Joint          Dessert Shop   
Downtown Toronto                  Café    Seafood Restaurant   

                 7th Most Common Venue 8th Most Common Venue  \
Borough                                                        
Downtown Toronto  Gym / Fitness Center    Mexican Restaurant   
Downtown Toronto           Pizza Place  Fast Food Restaurant   
Downtown Toronto                Bakery             BBQ Joint   
East Toronto                     Diner        Discount Store   
Downtown Toronto              Beer Bar                Bakery   

                 9th Most Common Venue 10th Most Common Venue  
Borough                                                        
Downtown Toronto               Theater         Breakfast Spot  
Downtown Toronto    Italian Restaurant       Ramen Restaurant  
Downtown Toronto                  Park                  Diner  
East Toronto                   Dog Run       Doner Restaurant  
Downtown Toronto        Farmers Market         Breakfast Spot

In [52]:
# Create map
latitude = 43.654260
longitude = -79.520999

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# add cluster color
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# addccolors for markers
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## CLUSTER MAP FOR VENUE LOCATIONS IN DOWNTOWN TORONTO